### ISSUE TITLE
hstack with quantities returns quantity with uninitialized unit

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#5100]
The following example shows that `hstack` does correctly return a quantity, but the unit is not initialized:

``` python
In [6]: import numpy as np

In [7]: from astropy import units as u

In [8]: x = [1,2] * u.m

In [9]: y = [3,4] * u.m

In [10]: np.hstack([x, y])
Out[10]: <Quantity [ 1., 2., 3., 4.] (Unit not initialised)>
```

cc @mhvk

**Comment by mhvk** [comment]
Right now, it is mentioned in the documentation as a [known issue](http://docs.astropy.org/en/latest/known_issues.html#quantity-issues). I think it is very annoying and wish it were easy to solve (but see the work-around using `u.Quantity([x, y]).ravel()`).

Just in case someone with a lot of energy comes along..: This goes wrong for all numpy stacking functions (which are all based no `concatenate`). In the `concatenate` code, first an array of the appropriate subtype for the results is created, but the relevant [call to `PyArray_NewFromDescr`](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/multiarraymodule.c#L422) does not update the unit -- I guess one would need to add code to call `__array_finalize__` somewhere. Next the different arrays are [copied using `PyArray_AssignArray`](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/multiarraymodule.c#L449) which I don't think would care about units at all: it would somehow need to go through `__setitem__` to ensure things would go right. Frankly, starting from scratch I would just write this as a python function, in which case this would be fairly easy.

**Comment by mhvk** [comment]
fixes by #8808
